azkaban-common/src/main/java/azkaban/executor/container/watch/FlowStatusManagerListener.java

```diff
@@ -239,6 +239,9 @@ private void updatePodStatus(AzPodStatusMetadata event) {
           "Flow execution-id %d for pod %s does not have a final status in database and will be "
               + "finalized.", executionId, event.getPodName()));
       final String reason = "Flow Pod execution was completed.";
+      if (!validateTransition(event)) {
```

Comment:
The method `compareAndFinalizeFlowStatus` should not be invoking `validateTransition` directly, would be better to let this method be narrow in scope which and thereby retain its reusability.  Also note that the check for `validateTransition` has already occurred before this method will be invoked and calling it again in this method could be a little confusing.

I recommend,

1. Add an additional argument `Status finalStatus` to all of these methods - `compareAndFinalizeFlow`, `finalizeFlowAndDeleteContainer`,  `ExecutionControllerUtils.finalizeFlow`. The default value for this argument can be `FAILED` if needed (using overloaded methods with this argument missing)
2. In the method `validateAndPreProcess`, once we have found that the transition is invalid then invoke `finalizeFlowAndDeleteContainer ` with the `finalStatus` argument set to `POD_FAILED`